File: android/vbt-app/app/src/main/java/com/vbt/app/data/repository/AuthRepository.kt
package com.vbt.app.data.repository

import com.vbt.app.data.local.PreferencesManager
import com.vbt.app.data.remote.ApiService
import com.vbt.app.data.remote.LoginRequest
import com.vbt.app.data.remote.RegisterRequest
import com.vbt.app.data.remote.UserDto
import kotlinx.coroutines.flow.Flow
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class AuthRepository @Inject constructor(
    private val api: ApiService,
    private val prefs: PreferencesManager
) {
    suspend fun login(username: String, password: String): Result<UserDto> {
        return try {
            val response = api.login(LoginRequest(username = username, password = password))
            if (response.isSuccessful) {
                val body = response.body()
                if (body != null) {
                    prefs.saveToken(body.accessToken)
                    prefs.saveUser(body.user.id, body.user.username, body.user.role, body.user.coachId)
                    Result.success(body.user)
                } else {
                    Result.failure(Exception("Empty response body"))
                }
            } else {
                Result.failure(Exception("Invalid credentials: ${response.code()}"))
            }
        } catch (e: Exception) {
            Result.failure(Exception("Connection error: ${e.message}", e))
        }
    }

    suspend fun register(username: String, password: String): Result<UserDto> {
        return try {
            val response = api.register(RegisterRequest(username = username, password = password))
            if (response.isSuccessful) {
                val body = response.body()
                if (body != null) {
                    prefs.saveToken(body.accessToken)
                    prefs.saveUser(body.user.id, body.user.username, body.user.role, body.user.coachId)
                    Result.success(body.user)
                } else {
                    Result.failure(Exception("Pusta odpowiedź serwera"))
                }
            } else {
                val msg = response.errorBody()?.string() ?: "Błąd rejestracji"
                Result.failure(Exception(msg))
            }
        } catch (e: Exception) {
            Result.failure(Exception("Błąd połączenia: ${e.message}", e))
        }
    }

    suspend fun logout() {
        prefs.clear()
    }

    fun getCurrentRole(): Flow<String?> = prefs.getRole()

    fun getCurrentUserId(): Flow<Int?> = prefs.getUserId()

    fun getCurrentUsername(): Flow<String?> = prefs.getUsername()

    fun isLoggedIn(): Flow<Boolean> = prefs.isLoggedIn()
}

File: ios/VBT/VBT/Core/Auth/AuthRepository.swift
import Foundation

enum UserRole: String {
    case coach, athlete
}

/// Port 1:1 z Android `data/repository/AuthRepository.kt`. `@Observable`, żeby
/// `RootView` mogło reagować na zmianę stanu zalogowania bez ręcznego bindowania.
@Observable
final class AuthRepository {
    private let api: APIClient
    private let keychain: KeychainStore
    private let prefs: UserPreferences

    private(set) var currentUser: UserDto?

    var isLoggedIn: Bool { keychain.readToken() != nil && currentUser != nil }
    var currentRole: UserRole? { prefs.role.flatMap(UserRole.init) }

    init(api: APIClient, keychain: KeychainStore, prefs: UserPreferences) {
        self.api = api
        self.keychain = keychain
        self.prefs = prefs
        restoreSession()
    }

    /// Po starcie aplikacji: jeśli mamy token i dane usera z poprzedniej sesji, odtwórz stan
    /// bez ponownego logowania (nie ma osobnego endpointu /users/me, więc bazujemy na cache).
    private func restoreSession() {
        guard keychain.readToken() != nil,
              let id = prefs.userId,
              let username = prefs.username,
              let role = prefs.role else { return }
        currentUser = UserDto(id: id, email: nil, username: username, role: role, isActive: true, coachId: prefs.coachId)
    }

    func login(username: String, password: String) async -> Result<UserDto, Error> {
        do {
            let response: LoginResponse = try await api.send(.login(LoginRequest(username: username, password: password)))
            persist(response)
            return .success(response.user)
        } catch {
            return .failure(error)
        }
    }

    func register(username: String, password: String) async -> Result<UserDto, Error> {
        do {
            let response: LoginResponse = try await api.send(.register(RegisterRequest(username: username, password: password)))
            persist(response)
            return .success(response.user)
        } catch {
            return .failure(error)
        }
    }

    func logout() {
        keychain.clear()
        prefs.clear()
        currentUser = nil
    }

    private func persist(_ response: LoginResponse) {
        keychain.saveToken(response.accessToken)
        prefs.save(userId: response.user.id, username: response.user.username, role: response.user.role, coachId: response.user.coachId)
        currentUser = response.user
    }
}
